### PR TITLE
Tx indexer

### PR DESCRIPTION
**Pull request by NateWilliams2**
Indexes transactions on the scrutinizer.

Store `TxReference`s containing each tx's height, blockHeight, and txIndex. 
Also provides the getTxByHeight api, allowing transactions to be queried by transaction height, not just block/txIndex

### CODE REVIEW

File: vochain/app.go [hunk 1]
```diff
@@ -350,6 +350,12 @@ func (app *BaseApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.
 			log.Debugf("rejected tx: %v", err)
 			return abcitypes.ResponseDeliverTx{Code: 1, Data: []byte(err.Error())}
 		}
+		for _, e := range app.State.eventListeners {
+			err := e.OnNewTx(app.State.height+1, uint32(app.State.TxCounter()))
```
Comment (p4u) [inline vochain/app.go]: use `app.Height()` instead

File: vochain/app.go [hunk 2]
```diff
@@ -350,6 +350,12 @@ func (app *BaseApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.
...
+			if err != nil {
```
Comment (p4u) [inline vochain/app.go]: `if err := e.OnNewTx()...`

File: vochain/app.go [hunk 3]
```diff
@@ -350,6 +350,12 @@ func (app *BaseApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.
...
+				log.Errorf("deliverTx: %v", err)
```
Comment (p4u) [inline vochain/app.go]: I think that should be a `Warnf` since the error is from a eventlistener

File: vochain/censusdownloader/censusdownloader.go [hunk 4]
```diff
@@ -81,6 +81,7 @@ func (c *CensusDownloader) OnProcess(pid, eid []byte, censusRoot, censusURI stri
 // NOT USED but required for implementing the interface
 func (c *CensusDownloader) OnCancel(pid []byte, txindex int32)                       {}
 func (c *CensusDownloader) OnVote(v *models.Vote, txindex int32)                     {}
+func (c *CensusDownloader) OnNewTx(blockHeight, txIndex uint32) error                { return nil }
```
Comment (p4u) [inline vochain/censusdownloader/censusdownloader.go]: I think its more consistent to use the same type for all `txIndex`. I think all of them should be uint32 but if you don't want to change so many code (its fine, we can do it afterwars in another commit) just use int32

File: vochain/scrutinizer/indexertypes/types.go [hunk 5]
```diff
@@ -12,6 +12,19 @@ import (
...
+const (
+	Entities     = 0
+	Processes    = 1
+	Envelopes    = 2
+	Transactions = 3
+)
+
+// CountStore holds the count of envelopes, processes, entities, or transactions
+type CountStore struct {
+	Type  int `badgerholdKey:"Type"`
```
Comment (p4u) [inline vochain/scrutinizer/indexertypes/types.go]: uint8 ?

File: vochain/scrutinizer/scrutinizer.go [hunk 6]
```diff
@@ -97,26 +97,44 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
 		return nil, err
 	}
 	startTime := time.Now()
-	envelopes, err := s.db.Count(&indexertypes.VoteReference{}, &badgerhold.Query{})
+
+	txCountStore := new(indexertypes.CountStore)
+	err = s.db.FindOne(txCountStore,
+		badgerhold.Where(badgerhold.Key).Eq(indexertypes.Transactions))
+	if err != nil {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go:103]: `if err = s.db...; err != nil {}`

File: vochain/scrutinizer/transaction.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	txReference := &indexertypes.TxReference{}
+	err := s.db.FindOne(txReference, badgerhold.Where(badgerhold.Key).Eq(height))
+	if err != nil {
+		return nil, fmt.Errorf("tx height %d not found: %v", height, err)
+	}
+	return txReference, nil
+}
+
+// OnNewTx stores the transaction reference in the indexer database
+func (s *Scrutinizer) OnNewTx(blockHeight, txIndex uint32) error {
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: We cannot store data on OnNewTx since its executed in deliverTx where the block can still be discarted. So you need to queue it and store the data on `Commit` instead

File: vochain/scrutinizer/transaction.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	txCount := atomic.AddUint64(&s.countTotalTransactions, 1)
+	err := s.db.Insert(txCount, &indexertypes.TxReference{
+		Index:        txCount,
+		BlockHeight:  blockHeight,
+		TxBlockIndex: txIndex,
+	})
+	if err != nil {
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: `if err := s.db.Insert...`

**Review by p4u — CHANGES_REQUESTED** [review]

**Comment by p4u** [comment]
@mvdan please also review

File: vochain/scrutinizer/scrutinizer_test.go [hunk 9]
```diff
@@ -953,3 +953,28 @@ func TestCountVotes(t *testing.T) {
...
+			qt.Assert(t, err, qt.IsNil)
+		}
+	}
+
+	for i := 0; i < 10; i++ {
+		for j := 0; j < 10; j++ {
+			ref, err := sc.GetTxReference(uint64(i*10 + j + 1))
+			qt.Assert(t, err, qt.IsNil)
+			qt.Assert(t, ref.BlockHeight, qt.CmpEquals(), uint32(i))
+			qt.Assert(t, ref.TxBlockIndex, qt.CmpEquals(), uint32(j))
```
Comment (mvdan) [inline vochain/scrutinizer/scrutinizer_test.go]: you can just use qt.Equals here, since these are just integers.

**Comment by NateWilliams2** [comment]
@p4u I addressed the comments. Pending is to change all tx indexes to uint32

File: vochain/scrutinizer/indexertypes/types.go [hunk 10]
```diff
@@ -12,6 +12,19 @@ import (
 	"go.vocdoni.io/proto/build/go/models"
 )
 
+const (
```
Comment (p4u) [inline vochain/scrutinizer/indexertypes/types.go:15]: These types should have a prefix and also a comment, else its hard to understand how to use them

File: vochain/scrutinizer/process.go [hunk 11]
```diff
@@ -153,20 +153,20 @@ func (s *Scrutinizer) ProcessList(entityID []byte,
 	return procs, err
 }
 
-// ProcessCount return the number of processes indexed
-func (s *Scrutinizer) ProcessCount(entityID []byte) int64 {
+// ProcessCount returns the number of processes indexed
+func (s *Scrutinizer) ProcessCount(entityID []byte) uint64 {
```
Comment (p4u) [inline vochain/scrutinizer/process.go:156]: We should try to unify integer types, I think for ProcessCount() uint32 should mot more than enough

File: vochain/scrutinizer/scrutinizer.go [hunk 12]
```diff
@@ -97,26 +99,40 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
 	}
 	startTime := time.Now()
-	envelopes, err := s.db.Count(&indexertypes.VoteReference{}, &badgerhold.Query{})
-	if err != nil {
-		return nil, fmt.Errorf("could not count the total envelopes: %w", err)
+
+	txCountStore := new(indexertypes.CountStore)
+	if err = s.db.FindOne(txCountStore,
+		badgerhold.Where(badgerhold.Key).Eq(indexertypes.Transactions)); err != nil {
+		log.Warnf("could not get the transaction count: %v", err)
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go:127]: If the counters cannot be recovered from the KV, we should calculate them (using badgerhold.Count()) and store

File: vochain/scrutinizer/scrutinizer.go [hunk 13]
```diff
@@ -97,26 +99,40 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
+		badgerhold.Where(badgerhold.Key).Eq(indexertypes.Processes)); err != nil {
+		log.Warnf("could not get the process count: %v", err)
 	}
+	entityCountStore := new(indexertypes.CountStore)
+	if err = s.db.FindOne(entityCountStore,
+		badgerhold.Where(badgerhold.Key).Eq(indexertypes.Entities)); err != nil {
+		log.Warnf("could not get the entity count: %v", err)
+	}
+
+	s.countTotalTransactions = txCountStore.Count
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: Since now we have the counters stored on the persistent KV, I don't think its worth to keep these in-memory caches.

File: vochain/scrutinizer/scrutinizer.go [hunk 14]
```diff
@@ -238,6 +254,16 @@ func (s *Scrutinizer) Commit(height uint32) error {
 		}
 	}
 
+	// Index new transactions
+	for _, tx := range s.newTxPool {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: It would be great if we can run this async, but we should be sure that the code is safe. Also I think we should abstract this code in a function (and just call something like `go updateTxIndex(s.newTxPool)`)

**Review by p4u — CHANGES_REQUESTED** [review]

File: vochain/scrutinizer/process.go (hunk 11 above)
Comment (NateWilliams2) [reply vochain/scrutinizer/process.go:156]: should I change the CountStore Count type to be uint32? This would include processes, entitites, envelopes, txs

File: vochain/scrutinizer/scrutinizer.go (hunk 13 above)
Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: Are you proposing to have a db transaction every time we want to access or update these values, i.e. every block?

File: vochain/scrutinizer/scrutinizer.go (hunk 14 above)
Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: I've updated the code to run async, and I've also tested it with the race flag a few times without any races.

Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: This update actually was causing many database transaction conflicts. Rather than handling them only for this case, I submitted https://github.com/vocdoni/vocdoni-node/pull/216 which handles transaction conflicts for most update functions. I've rebased this branch on that branch locally and verified that it fixes the problem.

The other issue with running this async is that there's no guarantee that txs will all be indexed after commit is called. For testing, this requires adding a time delay after calling Commit().

File: vochain/scrutinizer/process.go (hunk 11 above)
Comment (p4u) [reply vochain/scrutinizer/process.go:156]: You already changed it from int64 to uint64, no?

File: vochain/scrutinizer/process.go [hunk 15]
```diff
@@ -266,8 +273,15 @@ func (s *Scrutinizer) newEmptyProcess(pid []byte) error {
 	}
 	s.addVoteLock.Unlock()
 
-	// Increment the total process count cache
-	atomic.AddInt64(&s.countTotalProcesses, 1)
+	// Increment the total process count storage
+	s.db.UpdateMatching(&indexertypes.CountStore{}, badgerhold.Where(badgerhold.Key).Eq(indexertypes.CountStore_Processes), func(record interface{}) error {
```
Comment (p4u) [inline vochain/scrutinizer/process.go]: break lines

File: vochain/scrutinizer/scrutinizer.go [hunk 16]
```diff
@@ -97,33 +94,91 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
+				log.Warnf("could not store envelope count: %v", err)
+			}
+		}
+	}
+	processCountStore := new(indexertypes.CountStore)
+	if err = s.db.Get(indexertypes.CountStore_Processes, processCountStore); err != nil {
+		log.Warnf("could not get the process count: %v", err)
+		count, err := s.db.Count(&indexertypes.Process{}, &badgerhold.Query{})
+		if err != nil {
+			log.Warnf("could not count total processes: %v", err)
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: That should either return an error or end up in a Fatal

**Review by p4u — CHANGES_REQUESTED** [review]

File: vochain/scrutinizer/scrutinizer.go (hunk 14 above)
Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: Why do we have transaction conflicts here? I don't get it. We running the Inserts sync (not in parallel), so there should not be transaction conflicts.

Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: The conflict is actually on storing the transaction count. There aren't conflicts within each indexNewTxs call, but there could be multiple calls to this method running at the same time (ie. block 200 has 25 transactions and is indexing them asynchronously, then block 201 has only 1 transaction and indexes it much faster,  catching up with the previous call to indexNewTxs). The conflicts actually stop if I test and only update the txCount if there are txs in the block (rather than updating the txCount from x to x). We should do this anyways to avoid making unnecessary transactions, but the point is a race between transactions is possible here.

Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: This is also fixed by adding a mutex within this method, so only one block's transactions will be indexed at once

File: vochain/scrutinizer/process.go (hunk 11 above)
Comment (NateWilliams2) [reply vochain/scrutinizer/process.go:156]: yes, for processes. But now processes, entities, envelopes, and tx counts will all share the same type- CountStore. Do we want those all to be uint32?

File: vochain/scrutinizer/indexertypes/types.go [hunk 17]
```diff
@@ -12,6 +12,23 @@ import (
...
+	// CountStore_Entities is the key for the entity count db reference
+	CountStore_Entities uint8 = 0
```
Comment (p4u) [inline vochain/scrutinizer/indexertypes/types.go]: Use camel case please

File: vochain/scrutinizer/process.go (hunk 11 above)
Comment (p4u) [reply vochain/scrutinizer/process.go:156]: Better not make changes not related with this commit, so lets just change the minimum values. However uint32 makes sense in my opinion here

File: vochain/scrutinizer/process.go [hunk 18]
```diff
@@ -198,8 +201,12 @@ func (s *Scrutinizer) EntityProcessCount(entityId []byte) (uint32, error) {
 }
 
 // EntityCount return the number of entities indexed by the scrutinizer
-func (s *Scrutinizer) EntityCount() int64 {
-	return atomic.LoadInt64(&s.countTotalEntities)
+func (s *Scrutinizer) EntityCount() uint64 {
+	entityCountStore := &indexertypes.CountStore{}
+	if err := s.db.Get(indexertypes.CountStore_Entities, entityCountStore); err != nil {
+		log.Warnf("could not get the process count: %v", err)
```
Comment (p4u) [inline vochain/scrutinizer/process.go]: This should return 0

File: vochain/scrutinizer/scrutinizer.go [hunk 19]
```diff
@@ -97,33 +96,90 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
+// retrieveCounts returns a count for txs, envelopes, processes, and entities, in that order.
+// If no CountStore model is stored for the type, it counts all db entries of that type.
+func (s *Scrutinizer) retrieveCounts() (*indexertypes.CountStore,
+	*indexertypes.CountStore, *indexertypes.CountStore, *indexertypes.CountStore) {
+	var err error
+	txCountStore := new(indexertypes.CountStore)
+	if err = s.db.Get(indexertypes.CountStore_Transactions, txCountStore); err != nil {
+		log.Warnf("could not get the transaction count: %v", err)
+		count, err := s.db.Count(&indexertypes.TxReference{}, &badgerhold.Query{})
+		if err != nil {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go:129]: This method should return an Error and instead of printing it and keep the execution we should return the error (and stop execution).

File: vochain/scrutinizer/scrutinizer.go [hunk 20]
```diff
@@ -97,33 +96,90 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
+			log.Warnf("could not count total transactions: %v", err)
+		} else {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: No need for this else, instead we should check if the error is of type "key Not Found" (else return the error and stop the execution).

File: vochain/scrutinizer/scrutinizer.go [hunk 21]
```diff
@@ -271,11 +330,18 @@ func (s *Scrutinizer) Commit(height uint32) error {
...
 		})
 		wg.Wait()
-		s.indexTxLock.Unlock()
+		s.voteTxLock.Unlock()
 		log.Infof("indexed %d new envelopes, took %s",
 			len(s.voteIndexPool), time.Since(startTime))
-		// Add the envelopes to the countTotalEnvelopes var
-		atomic.AddUint64(&s.countTotalEnvelopes, uint64(len(s.voteIndexPool)))
+		envelopeCount := &indexertypes.CountStore{}
+		if err := s.db.Get(indexertypes.CountStore_Envelopes, envelopeCount); err != nil {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: Cannot we assume this index exist already (was created by scrutinizer initialization) and avoid doing two db operations here?

File: vochain/scrutinizer/vote.go [hunk 22]
```diff
@@ -218,7 +218,11 @@ func (s *Scrutinizer) GetEnvelopeHeight(processID []byte) (uint64, error) {
 	defer func() { log.Debugf("GetEnvelopeHeight took %s", time.Since(startTime)) }()
 	if len(processID) == 0 {
 		// If no processID is provided, count all envelopes
-		return atomic.LoadUint64(&s.countTotalEnvelopes), nil
+		envelopeCountStore := &indexertypes.CountStore{}
+		if err := s.db.Get(indexertypes.CountStore_Envelopes, envelopeCountStore); err != nil {
+			return envelopeCountStore.Count, err
```
Comment (p4u) [inline vochain/scrutinizer/vote.go]: If error return 0 not the value

**Review by p4u — CHANGES_REQUESTED** [review]

File: vochain/scrutinizer/scrutinizer.go (hunk 21 above)
Comment (NateWilliams2) [reply vochain/scrutinizer/scrutinizer.go]: You mean by using updateMatching instead?

Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: Yes

File: vochain/scrutinizer/indexertypes/types.go [hunk 23]
```diff
@@ -12,6 +12,23 @@ import (
 	"go.vocdoni.io/proto/build/go/models"
 )
 
+const (
+	// CountStoreEntities is the key for the entity count db reference
```
Comment (p4u) [inline vochain/scrutinizer/indexertypes/types.go:16]: Use `iota`

File: vochain/scrutinizer/process.go [hunk 24]
```diff
@@ -153,20 +152,24 @@ func (s *Scrutinizer) ProcessList(entityID []byte,
...
 	defer func() { log.Debugf("ProcessCount took %s", time.Since(startTime)) }()
 	var c uint32
 	var err error
 	if len(entityID) == 0 {
-		// If no entity ID, return the cached count of all processes
-		return atomic.LoadInt64(&s.countTotalProcesses)
+		// If no entity ID, return the stored count of all processes
+		processCountStore := &indexertypes.CountStore{}
+		if err = s.db.Get(indexertypes.CountStoreProcesses, processCountStore); err != nil {
+			log.Warnf("could not get the process count: %v", err)
```
Comment (p4u) [inline vochain/scrutinizer/process.go]: That should be an `Error()` since its something that should never happen (actually it should be a Panic() but in general we are not making the components Panic)

File: vochain/scrutinizer/scrutinizer.go [hunk 25]
```diff
@@ -97,33 +98,109 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
 	s.App.State.AddEventListener(s)
 	s.envelopeHeightCache = lru.New(countEnvelopeCacheSize)
 	s.resultsCache = lru.New(resultsCacheSize)
 	return s, nil
 }
 
+// retrieveCounts returns a count for txs, envelopes, processes, and entities, in that order.
+// If no CountStore model is stored for the type, it counts all db entries of that type.
+func (s *Scrutinizer) retrieveCounts() (*indexertypes.CountStore,
+	*indexertypes.CountStore, *indexertypes.CountStore, *indexertypes.CountStore, error) {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: Wouldn't it make more sense to return a `map[uint8]uint64` ?

File: vochain/scrutinizer/scrutinizer.go [hunk 26]
```diff
@@ -97,33 +98,109 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication) (*Scrutinizer,
...
+		txCountStore.Type = indexertypes.CountStoreTransactions
+		if err := s.db.Upsert(txCountStore.Type, txCountStore); err != nil {
+			return nil, nil, nil, nil, fmt.Errorf("could not store transaction count: %v", err)
+		}
+	}
+	envelopeCountStore := new(indexertypes.CountStore)
+	if err = s.db.Get(indexertypes.CountStoreEnvelopes, envelopeCountStore); err != nil {
+		log.Warnf("could not get the envelope count: %v", err)
+		count, err := s.db.Count(&indexertypes.VoteReference{}, &badgerhold.Query{})
+		if err != nil {
```
Comment (p4u) [inline vochain/scrutinizer/scrutinizer.go]: count is 0 in case of error (zero-value for uint64) so you can simplify this logic:

```if err != nil &&  err != badger.ErrKeyNotFound { return .... }```

Comment (p4u) [reply vochain/scrutinizer/scrutinizer.go]: Same on all cases.

File: vochain/scrutinizer/transaction.go [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+	s.newTxPool = append(s.newTxPool, &indexertypes.TxReference{
+		BlockHeight:  blockHeight,
+		TxBlockIndex: txIndex,
+	})
+}
+
+// indexNewTxs indexes the txs pending in the newTxPool and updates the transaction count
+// this function should only be called within Commit(), on a new block.
+func (s *Scrutinizer) indexNewTxs(txList []*indexertypes.TxReference) {
+	if len(txList) > 0 {
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: `if len(txList) == 0 { return }`

File: vochain/scrutinizer/transaction.go [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+		s.txIndexLock.Lock()
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: use defer for Unlock()

File: vochain/scrutinizer/transaction.go [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+		txCount := &indexertypes.CountStore{}
+		if err := s.db.Get(indexertypes.CountStoreTransactions, txCount); err != nil {
+			log.Errorf("could not get tx count: %v", err)
```
Comment (p4u) [inline vochain/scrutinizer/transaction.go]: Are we sure we want to keep the execution in case of error??

Please review all this cases of your PR, either "return" or "continue" or something is required if a query returns an error.

**Review by p4u — CHANGES_REQUESTED** [review]

**Comment by NateWilliams2** [comment]
addressed most recent comments @p4u

**Review by p4u — APPROVED** [review]
